packages/babel-generator/src/node/parentheses.js

```diff
@@ -35,14 +35,21 @@ export function NullableTypeAnnotation(node: Object, parent: Object): boolean {
   return t.isArrayTypeAnnotation(parent);
 }
 
-export function FunctionTypeAnnotation(node: Object, parent: Object): boolean {
+export function FunctionTypeAnnotation(
+  node: Object,
+  parent: Object,
+  printStack: Array<Object>,
+): boolean {
   return (
     // (() => A) | (() => B)
     t.isUnionTypeAnnotation(parent) ||
     // (() => A) & (() => B)
     t.isIntersectionTypeAnnotation(parent) ||
     // (() => A)[]
-    t.isArrayTypeAnnotation(parent)
+    t.isArrayTypeAnnotation(parent) ||
+    // <T>(A: T): (T => T[]) => B => [A, B]
+    (t.isTypeAnnotation(parent) &&
+      t.isArrowFunctionExpression(printStack[printStack.length - 3]))
```

Comment:
What does `printStack[printStack.length - 3]`...... mean?

I assume this is related to Flow emit; the example syntax looks like invalid TypeScript.

Also, would the structure be the same if `<T>(A: T): // ...` (declaration signature) was written as `<T>(A: T) => /...` (expression signature)?

Reply:
Yeah, it's Flow emit related. I can add a comment, the printStack bit is just checking the grandparent (since these are only passed node, its parent, and the whole stack).